### ISSUE TITLE
Rendring need improvments 

### ISSUE THREAD

**Issue by rickardjanson** [dfilatov/vidom#46]
Hi. I notice the diff/ patch now starts to get good performance. But still slow comparing to the fastest implementations. 

*render* in the benchmark are sky high. 450 in average on my computer. Inferno and cito are around 200.

This is a huge performance issues. How you fix that?

**Comment by dfilatov** [comment]
I'm not sure about "render in the benchmark are sky high" but I'm ready to hear any meaningful advice.

**Comment by rickardjanson** [comment]
When you stated in your readme that you aim to be the fastest one, and the fastest one are around 200 and yours areound 460 - 510. There is a difference around 57% 

You have to re-think the use of prototype you are using, but as I tried to show you earlier in an code I tried to send  you, you could set a "rootNode". This rootNode is the only one who create a new TagNode, and not the others etc.

Else there is not much you can do then tweaking. And optimizing.

**Comment by rickardjanson** [comment]
@dfilatov A dirty trick could be to try to avoid re-rendring the nodes, if they already are rendered. And re-use them instead

**Comment by dfilatov** [comment]
Why do you think that the nodes are rerendered? The nodes are rendered inside engine (and aren't rerendered if they are already rendered), user doesn't interact directly with them.
Maybe I missed something and you have case when the nodes are rerendered?

**Comment by rickardjanson** [comment]
@dfilatov They are not rerendered in the way you think I'm thinking. But to understand this, you need to understand some low level architecture. I don't think I'm going to help you with that answer or how to do that because I feel this start to be vaste of my time.

What I can say to you, and promise you, is that you have a couple of flaws - huge mistakes - in how you do this concept. And if you listen to my advice here without any reference (*ofc I can give references*) your rendring will be improved in Internet Explorer, and also fall down to around 372 on my computer,comparing with today your's are now 460+ on my computer.

You do it easy. And just like this.

1. Create the child nodes first
2. Attach the nodes to it's container
3. Render the node

Got it?

So. Have a look at this lines: https://github.com/dfilatov/vidom/blob/master/lib/nodes/TagNode.js#L58-L65

**Only** create the node and attach the node to the parent container first. THEN, if you render the other nodes, that is not attached to any parent container, you do nothing. So you need to **only** attach the child nodes to - finished rendered - on it's parent.

Say you got 400 nodes. Only 1 parent here, right?  So why use 399 CPU cycles to attach to a parent container, when you can attach them to a virtual node instead and then when you got your chain of child nodes finished rendered, you attach 1 time to the parent container.

You only got parent container, but you can have virtual nodes to hook things onto.

I will not go more in depth on this, I just gave you the idea.

**Comment by dfilatov** [comment]
Sorry, It's really difficult to me to understand all mess you have written.

For instance:
> when you got your chain of child nodes finished rendered, you attach 1 time to the parent container.

How do you attach a chain of child DOM nodes only 1 time? Using `DocumentFragment`?

Ps. If you feel you waste your time you'd better stop.

**Comment by rickardjanson** [comment]
I'm going to stop. So to end this, read this post regarding what I mentioned regarding IE. 
https://github.com/localvoid/vdom-benchmark/issues/15#issuecomment-71689398

For the "mess I wrote", well, 

Look here. IF Internet Explorer, the things mentioned in the post are done. https://github.com/joelrich/citojs/blob/master/dist/cito.js#L258-L259

Then the child nodes are rendered as I also described without attched to the parent container.

Have a look here:  https://github.com/joelrich/citojs/blob/master/dist/cito.js#L265

After all nodes are rendered, they are set on the parent container IF it's the root node. If not, it will not happend. Probably here my mess come in?

Look here:  https://github.com/joelrich/citojs/blob/master/dist/cito.js#L286-L288

So study that code, and compare with what I wrote. I can also OFC point to other sources etc.
